src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -157,18 +160,53 @@ public void arcadeDrive(double throttle, double turnModifier){
+   */
   public void curvatureDrive(double throttle, double curvature) {
-    if(throttle != 0.0 && curvature == 0.0 && !gyroDisabled) { //driving straight and no turn
-      if(this.desiredAngle == Integer.MAX_VALUE) { //means robot just started driving straight
-        this.desiredAngle = gyroscope.getGyroAngle(); 
-      }
-      curvature = this.getAngularError(desiredAngle) * P_VALUE; 
-      this.setLeftPower(throttle - curvature);
-      this.setRightPower(throttle + curvature);
-    }
-    else { // when robot isn't driving straight
-      this.desiredAngle = Integer.MAX_VALUE; //if turn is greater than 0 or if robot is still
-      differentialDrive.curvatureDrive(-throttle, curvature, (Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE));
+
+    if(Math.abs(throttle) < QUICK_TURN_THROTTLE_DEADZONE){
+      // If throttle is less than the deadzone, treat inputs as arcade
+      this.arcadeDrive(throttle, curvature);
+    }else{
+      // Otherwise, treat inputs as curvature
+
+      // Calculate ideal left and right speeds based on curvature
+      WheelSpeeds desiredSpeeds = DifferentialDrive.curvatureDriveIK(throttle, curvature, false);
+
+      // To execute closed loop and correct unwanted turning, we need to directly
+      // compare a desired quanity to the current quantity to get an error.
+      // The units on desired speeds are hard to relate to turning units directly,
+      // so we need a way to cancel them out. If we control the curvature rate instead (which
+      // is turn rate / linear rate), we can cancel out some of the units into degrees/meter,
+      // making comparison easier.
+
+      // Desired curvature rate (angular speed / linear speed)
+      double desiredTurnRate = (180 / Math.PI) * (desiredSpeeds.left - desiredSpeeds.right) / (Constants.TRACK_WIDTH_METERS / 2);
+      double desiredSpeed = Math.abs(desiredSpeeds.left + desiredSpeeds.right) / 2.0;
+      // account for the possiblity of dividing by zero by setting miniumum "desired speed" value
+      double desiredCurvature = desiredTurnRate / Math.max(desiredSpeed, QUICK_TURN_THROTTLE_DEADZONE);
+
+      // Measured curvature rate (angular speed / linear speed)
+      double measuredTurnRate = this.gyroscope.getRate();
+      double measuredSpeed = Math.abs(this.leftGearbox.getEncoderRate() + this.rightGearbox.getEncoderRate()) / 2.0;
+      // account for the possiblity of dividing by zero. Anything less than 5cm/s or so is too noisy, set floor
+      double measuredCurvature = measuredTurnRate/Math.max(measuredSpeed, 0.05);
+
+      // Modify wheel speeds based on curvature error
+      double turnRateCorrection = (desiredCurvature - measuredCurvature) * kP_CURVATURE_DRIVE;
+      desiredSpeeds.left += turnRateCorrection;
+      desiredSpeeds.right -= turnRateCorrection;
+
+      // send the modified wheel speeds to the motors
+      differentialDrive.tankDrive(desiredSpeeds.left, desiredSpeeds.right);
```

Comment:
I think we could possibly pull out the SimpleMotorFeedForward and the PIDController from the trajectory stuff and use them here. I'm not super confident about this assessment though. The P in the PIDController should be relevant here. We would then add the PID value + the FF value and set the tankDriveVolts

Reply:
Pulling in your comment [here](https://github.com/PearceRobotics/rapid-react-2022/pull/49/files#r807453750): combining these two thoughts, what if we made a closed loop tank drive method using velocities? Pretty much every differential drive closed loop problem can boil down to doing a loop on velocity, including this one.

Reply:
Ideally, if we want to do this a student would take over.

Reply:
I'm happy to use velocity based closed loop control. We can get this merged and roll that feature into a new ticket for a student to work on.